### ISSUE TITLE
Shouldn't .toFirebaseValue() return a Map<String, Object>?

### ISSUE THREAD

**Issue by renanferrari** [mattlogan/auto-value-firebase#13]
If `.toFirebaseValue()` returned a `Map<String, Object>` instead of an `Object` we would be able to use `ref.updateChildren(myObject.toFirebaseValue())`.

Is there any specific reason behind the decision of returning an `Object` that would prevent this? Am I missing something?

**Comment by mattlogan** [comment]
This extension doesn't generate a `toFirebaseValue()` method. The `toFirebaseValue()` method that you see in the README was added by me in the `AutoValue` class that I defined (`Taco` in this example).

The existing mechanism for converting from `AutoValue` objects to objects that Firebase will recognize is via the generated constructor in the generated `FirebaseValue` class. This lets you pass the `FirebaseValue` object as an argument to `DatabaseReference#setValue(Object)`.

You are free to use this constructor in a `toFirebaseValue()` implementation, but you can define this however you want. You could also add a method called `toMap()` on each of your AutoValue classes that returns a map, like this:

```java
@AutoValue @FirebaseValue
public abstract class Taco {

  public static Taco create(String name, List<Ingredient> ingredients, Review review, Status status) {
    return new AutoValue_Taco(name, ingredients, review, status);
  }

  public static Taco create(DataSnapshot dataSnapshot) {
    return dataSnapshot.getValue(AutoValue_Taco.FirebaseValue.class).toAutoValue();
  }

  public Object toFirebaseValue() {
    return new AutoValue_Taco.FirebaseValue(this);
  }

  public Map<String, Object> toMap() {
    Map<String, Object> map = new HashMap<>();
    map.put("name", name());
    List<Map<String, Object>> ingredients = new ArrayList<>();
    for (Ingredient ingredient : ingredients()) {
      ingredients.add(ingredient.toMap());
    }
    map.put("rating", review().toMap());
    return map;
  }

  public abstract String name();

  public abstract List<Ingredient> ingredients();

  public abstract Review review();

  @FirebaseAdapter(StatusAdapter.class)
  public abstract Status status();
}
```

I might look into a way to generate this `toMap()` method so you don't have to write it yourself, but I don't see it replacing the `FirebaseValue` constructor as the primary mechanism for converting your `AutoValue` objects.

**Comment by renanferrari** [comment]
@mattlogan Thanks for the answer! Yeah, generating this method would be perfect. And you're totally right about the constructor.

**Comment by chetbox** [comment]
We've found that converting to `Map<String, Object>` is necessary to use [`DatabaseReference.updateChildren`](https://firebase.google.com/docs/reference/android/com/google/firebase/database/DatabaseReference.html#updateChildren(java.util.Map%3Cjava.lang.String%2C%20java.lang.Object%3E)) as well. (In our case we want to use "update" rather "save" to prevent values written by other clients being removed if they're not specified in out `@FirebaseValue` class.)

[This AutoValue extension](https://github.com/cynnyx/auto-value-map) will implment a `toMap()` method for you but it doesn't work for nested AutoValue classes.

I think it would be neat to include a `Map<String, Object>` option in auto-value-firebase given that it can be used for `DatabaseReference.updateChildren`. Could we make `FirebaseValue` implement `Map<String, Object>` or provide a `toMap()` method?

Edit: Grammar. Also looking back at this I've repeated a lot what's been said here already.
